File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/data/im/ImEventDecoder.kt
package com.tabtin.mobile.data.im

import com.tabtin.mobile.data.api.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject

/**
 * TabChat 实时事件（Centrifugo publication 解析结果）。
 *
 * 逐字段对齐后端 `IMOutboxService.enqueue` 的 payload 与 Electron `useCentrifugoClient.ts`。
 * 未识别 / 解析失败返回 [Unknown] 或 null，调用方据此丢弃脏包（不影响连接）。
 */
public sealed interface ImRealtimeEvent {
    /** `chat:{conv}` 新消息。 */
    public data class Message(val message: ImMessage) : ImRealtimeEvent
    /** `chat:{conv}` 消息被编辑（携带完整新消息）。 */
    public data class MessageEdited(val message: ImMessage) : ImRealtimeEvent
    /** `chat:{conv}` 消息被撤回（软删）。 */
    public data class MessageDeleted(val messageId: Int) : ImRealtimeEvent
    public data class MessagePinned(val message: ImMessage) : ImRealtimeEvent
    public data class MessageUnpinned(val messageId: Int) : ImRealtimeEvent
    /** `chat:{conv}` 表情回应增删（[added]=false 为移除）。 */
    public data class Reaction(
        val messageId: Int,
        val userId: String,
        val emoji: String,
        val added: Boolean,
    ) : ImRealtimeEvent
    /** `chat:{conv}` 已读回执：某用户已读到 [lastReadSeq]。 */
    public data class ReadReceipt(val payload: ImReadReceiptEvent) : ImRealtimeEvent
    /** `personal:{userId}` 未读更新（会话列表角标即时刷新）。 */
    public data class UnreadUpdate(val payload: ImUnreadUpdate) : ImRealtimeEvent
    /** `personal:{userId}` 新会话（如新建 DM），携带会话摘要（同列表项形状）。 */
    public data class ConversationNew(val conversation: ImConversation) : ImRealtimeEvent
    /** 最后一条消息被编辑或撤回；只刷新目录摘要，不增加未读。 */
    public data class ConversationPreviewUpdated(val payload: ImConversationPreviewUpdate) : ImRealtimeEvent
    /** 当前用户给某会话贴/撕标签后的权威快照。 */
    public data class ConversationLabelsUpdated(val payload: ImConversationLabelsUpdatedEvent) : ImRealtimeEvent
    /** 共享会话参与者资料变化；私聊目录就地更新，活动群详情按版本重拉。 */
    public data class UserProfileUpdated(val profile: ImUserProfileUpdated) : ImRealtimeEvent
    /** 会话资料或成员变化；调用方重拉详情与目录。 */
    public data object ConversationChanged : ImRealtimeEvent
    /** `chat:{conv}` 对端正在输入。 */
    public data class Typing(val userId: String) : ImRealtimeEvent
    public data class HandoffUpdate(val handoffId: String) : ImRealtimeEvent
    public data class SessionShareUpdate(val shareId: String) : ImRealtimeEvent
    public data class AgentMessageStream(val payload: ImAgentMessageStreamEvent) : ImRealtimeEvent
    public data class AgentMessageFinal(val payload: ImAgentMessageFinalEvent) : ImRealtimeEvent
    public data class AgentMessageError(val payload: ImAgentMessageErrorEvent) : ImRealtimeEvent
    public data class AiError(val agentName: String, val reason: String) : ImRealtimeEvent
    public data class AiSuggestTask(
        val conversationId: String?,
        val messageId: Int?,
        val agentName: String,
    ) : ImRealtimeEvent
    /** 本期未处理的事件类型（保留 type 便于日志/扩展）。 */
    public data class Unknown(val type: String) : ImRealtimeEvent
}

@Serializable
public data class ImAgentMessageStreamEvent(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("message_ref") val messageRef: String,
    @SerialName("agent_session_ref") val agentSessionRef: String,
    @SerialName("sender_id") val senderId: String,
    @SerialName("sender_name") val senderName: String = "",
    @SerialName("sender_avatar") val senderAvatar: String = "",
    val delta: String,
    @SerialName("stream_seq") val streamSeq: Int,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
public data class ImAgentMessageFinalEvent(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("message_ref") val messageRef: String,
    @SerialName("agent_session_ref") val agentSessionRef: String,
    @SerialName("sender_id") val senderId: String,
    @SerialName("sender_name") val senderName: String = "",
    @SerialName("sender_avatar") val senderAvatar: String = "",
    val content: String = "",
    @SerialName("message_type") val messageType: Int = ImMessageType.TEXT,
    val metadata: ImMessageMetadata? = null,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
public data class ImAgentMessageErrorEvent(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("message_ref") val messageRef: String,
    @SerialName("agent_session_ref") val agentSessionRef: String,
    @SerialName("sender_id") val senderId: String,
    @SerialName("sender_name") val senderName: String = "",
    @SerialName("sender_avatar") val senderAvatar: String = "",
)

/** `im.read.receipt` 负载。 */
@Serializable
public data class ImReadReceiptEvent(
    @SerialName("conversation_id") val conversationId: String = "",
    @SerialName("user_id") val userId: String = "",
    @SerialName("last_read_message_id") val lastReadMessageId: Int? = null,
    @SerialName("last_read_seq") val lastReadSeq: Int = 0,
)

/** `im.reaction.added` / `im.reaction.removed` 负载。 */
@Serializable
public data class ImReactionEvent(
    @SerialName("message_id") val messageId: Int = 0,
    @SerialName("user_id") val userId: String = "",
    val emoji: String = "",
)

/** `im.message.deleted` 负载。 */
@Serializable
public data class ImMessageDeletedEvent(
    @SerialName("message_id") val messageId: Int = 0,
)

@Serializable
private data class ImHandoffUpdateEvent(
    @SerialName("handoff_id") val handoffId: String = "",
)

@Serializable
private data class ImSessionShareUpdateEvent(
    @SerialName("share_id") val shareId: String = "",
)

@Serializable
private data class ImAiErrorEvent(
    @SerialName("agent_name") val agentName: String = "",
    val reason: String = "",
)

@Serializable
private data class ImAiSuggestTaskEvent(
    @SerialName("conversation_id") val conversationId: String? = null,
    @SerialName("message_id") val messageId: Int? = null,
    @SerialName("agent_name") val agentName: String = "",
)

/**
 * `im.unread.update` 负载：两种形状共用同一 type——
 * 1) 新消息：带 `message_id` / `preview` 等（personal_base + mention）；
 * 2) 已读回写：带 `marked_read` / `last_read_seq`（本人 mark_as_read 后回推给自己）。
 */
@Serializable
public data class ImUnreadUpdate(
    @SerialName("conversation_id") val conversationId: String = "",
    @SerialName("organization_id") val organizationId: String = "",
    @SerialName("directory_scope_id") val directoryScopeId: String? = null,
    @SerialName("message_id") val messageId: Int = 0,
    @SerialName("message_seq") val messageSeq: Int = 0,
    @SerialName("sender_id") val senderId: String = "",
    @SerialName("sender_name") val senderName: String = "",
    val preview: String = "",
    @SerialName("last_message_at") val lastMessageAt: String? = null,
    val mention: Boolean = false,
    /** 非 null 表示这是「已读回写」事件（应把该会话未读清零），与新消息事件互斥。 */
    @SerialName("marked_read") val markedRead: Int? = null,
) {
    public val isMarkedReadEvent: Boolean get() = markedRead != null
}

@Serializable
public data class ImConversationLabelsUpdatedEvent(
    @SerialName("conversation_id") val conversationId: String = "",
    val labels: List<ImConversationLabel> = emptyList(),
)

@Serializable
public data class ImConversationPreviewUpdate(
    @SerialName("conversation_id") val conversationId: String = "",
    @SerialName("organization_id") val organizationId: String = "",
    @SerialName("directory_scope_id") val directoryScopeId: String? = null,
    @SerialName("message_id") val messageId: Int = 0,
    @SerialName("message_seq") val messageSeq: Int = 0,
    val preview: String = "",
    @SerialName("last_message_at") val lastMessageAt: String? = null,
)

/** `im.user.profile.updated` 负载。 */
@Serializable
public data class ImUserProfileUpdated(
    @SerialName("id") val userId: String = "",
    val nickname: String = "",
    val username: String = "",
    val avatar: String = "",
    @SerialName("avatar_version") val avatarVersion: String = "",
    val revision: Int = 0,
) {
    public val displayName: String
        get() = nickname.trim().ifEmpty { username.trim() }
}

/**
 * Centrifugo publication 原始字节 → 类型化事件。
 *
 * 外层信封形如 `{ "type": "im.*", "event_id": "...", "data": {...} }`；`im.typing` 例外，
 * 字段直接铺在顶层：`{ "type": "im.typing", "user_id": "..." }`。
 */
public object ImEventDecoder {

    public fun decode(raw: ByteArray): ImRealtimeEvent? =
        decode(raw.toString(Charsets.UTF_8))

    public fun decode(text: String): ImRealtimeEvent? {
        val root: JsonObject = runCatching { json.parseToJsonElement(text).jsonObject }
            .getOrNull() ?: return null
        val type = (root["type"] as? JsonPrimitive)
            ?.takeIf { it.isString }
            ?.contentOrNull ?: return null
        return when (type) {
            "im.message" -> root.decodeData<ImMessage>()?.let { ImRealtimeEvent.Message(it) }
            "im.message.edited" -> root.decodeData<ImMessage>()?.let { ImRealtimeEvent.MessageEdited(it) }
            "im.message.deleted" ->
                root.decodeData<ImMessageDeletedEvent>()?.let { ImRealtimeEvent.MessageDeleted(it.messageId) }
            "im.message.pinned" -> root.decodeData<ImMessage>()?.let { ImRealtimeEvent.MessagePinned(it) }
            "im.message.unpinned" ->
                root.decodeData<ImMessageDeletedEvent>()?.let { ImRealtimeEvent.MessageUnpinned(it.messageId) }
            "im.reaction.added", "im.reaction.removed" ->
                root.decodeData<ImReactionEvent>()?.let {
                    ImRealtimeEvent.Reaction(
                        messageId = it.messageId,
                        userId = it.userId,
                        emoji = it.emoji,
                        added = type == "im.reaction.added",
                    )
                }
            "im.read.receipt" -> root.decodeData<ImReadReceiptEvent>()?.let { ImRealtimeEvent.ReadReceipt(it) }
            "im.unread.update" -> root.decodeData<ImUnreadUpdate>()?.let { ImRealtimeEvent.UnreadUpdate(it) }
            "im.conversation.new" -> root.decodeData<ImConversation>()?.let { ImRealtimeEvent.ConversationNew(it) }
            "im.conversation.preview.updated" -> root.decodeData<ImConversationPreviewUpdate>()
                ?.let { ImRealtimeEvent.ConversationPreviewUpdated(it) }
            "im.conversation.labels.updated" -> root.decodeData<ImConversationLabelsUpdatedEvent>()
                ?.let { ImRealtimeEvent.ConversationLabelsUpdated(it) }
            "im.user.profile.updated" -> root.decodeData<ImUserProfileUpdated>()
                ?.let { ImRealtimeEvent.UserProfileUpdated(it) }
            "im.conversation.updated", "im.member.joined", "im.member.left" ->
                ImRealtimeEvent.ConversationChanged
            // typing 顶层平铺（无 data 信封）。
            "im.typing" -> {
                val userId = (root["user_id"] as? JsonPrimitive)
                    ?.takeIf { it.isString }
                    ?.contentOrNull ?: return null
                ImRealtimeEvent.Typing(userId)
            }
            "im.handoff.update" -> root.decodeData<ImHandoffUpdateEvent>()
                ?.takeIf { it.handoffId.isNotBlank() }
                ?.let { ImRealtimeEvent.HandoffUpdate(it.handoffId) }
            "im.session_share.update" -> root.decodeData<ImSessionShareUpdateEvent>()
                ?.takeIf { it.shareId.isNotBlank() }
                ?.let { ImRealtimeEvent.SessionShareUpdate(it.shareId) }
            "im.agent.message.stream" -> root.decodeData<ImAgentMessageStreamEvent>()
                ?.let { ImRealtimeEvent.AgentMessageStream(it) }
            "im.agent.message.final" -> root.decodeData<ImAgentMessageFinalEvent>()
                ?.let { ImRealtimeEvent.AgentMessageFinal(it) }
            "im.agent.message.error" -> root.decodeData<ImAgentMessageErrorEvent>()
                ?.let { ImRealtimeEvent.AgentMessageError(it) }
            "im.ai.error" -> root.decodeData<ImAiErrorEvent>()
                ?.let { ImRealtimeEvent.AiError(it.agentName, it.reason) }
            "im.ai.suggest_task" -> root.decodeData<ImAiSuggestTaskEvent>()
                ?.let { ImRealtimeEvent.AiSuggestTask(it.conversationId, it.messageId, it.agentName) }
            else -> ImRealtimeEvent.Unknown(type)
        }
    }

    /** 从信封的 `data` 子对象解码为 [T]；缺 `data` 或类型异常返回 null（丢弃脏包）。 */
    private inline fun <reified T> JsonObject.decodeData(): T? {
        val data: JsonElement = this["data"] ?: return null
        return runCatching { json.decodeFromJsonElement<T>(data) }.getOrNull()
    }
}

File: apps/tabtin-ios/Tabtin/Core/IM/IMEventDecoder.swift
import Foundation

/// TabChat 实时事件（Centrifugo publication 解析结果）。
///
/// 覆盖消息、会话、已读、卡片状态和个人 Agent 提示；无法识别的新增事件保留 type，
/// 逐字段对齐 Electron `useCentrifugoClient.ts` 与后端 outbox payload。
enum IMRealtimeEvent: Sendable, Equatable {
    /// `chat:{conv}` 新消息，`data` 为完整 `IMMessage`。
    case message(IMMessage)
    /// `chat:{conv}` 消息被编辑，`data` 为完整（新）`IMMessage`。
    case messageEdited(IMMessage)
    /// `chat:{conv}` 消息被撤回（软删）。
    case messageDeleted(messageId: Int)
    /// `chat:{conv}` 消息被置顶，负载为完整消息。
    case messagePinned(IMMessage)
    /// `chat:{conv}` 消息被取消置顶。
    case messageUnpinned(messageId: Int)
    /// `chat:{conv}` 表情回应增删（`added=false` 为移除）。
    case reaction(messageId: Int, userId: String, emoji: String, added: Bool)
    /// `chat:{conv}` 已读回执：某用户已读到 `lastReadSeq`。
    case readReceipt(IMReadReceiptEvent)
    /// `personal:{userId}` 未读更新（前台会话列表角标即时刷新）。
    case unreadUpdate(IMUnreadUpdate)
    /// `personal:{userId}` 新会话（如新建 DM），`data` 为会话摘要（同列表项形状）。
    case conversationNew(IMConversation)
    /// 最后一条消息被编辑或撤回；只刷新目录摘要，不增加未读。
    case conversationPreviewUpdated(IMConversationPreviewUpdate)
    /// 当前用户给某会话贴/撕标签后的权威快照。
    case conversationLabelsUpdated(conversationId: String, labels: [IMConversationLabel])
    /// 共享会话参与者的资料发生变化；私聊目录就地更新，活动群详情按版本重拉。
    case userProfileUpdated(IMUserProfileUpdated)
    /// 会话资料或成员发生变化；详情和目录应重新拉取权威快照。
    case conversationChanged
    /// `chat:{conv}` 对端正在输入。
    case typing(userId: String)
    /// 交接包状态变化；卡片收到后按 id 重拉详情，不能信任事件快照替代权限校验。
    case handoffUpdate(handoffId: String)
    /// 任务共享状态变化；卡片按 share id 失效缓存并重拉详情。
    case sessionShareUpdate(shareId: String)
    /// Agent 可见文本增量；按 `messageRef` 聚合，`streamSeq` 负责去重与乱序保护。
    case agentMessageStream(IMAgentMessageStreamEvent)
    /// Agent 最终可见消息；替换同 `messageRef` 的临时流式消息。
    case agentMessageFinal(IMAgentMessageFinalEvent)
    /// Agent 执行失败；清理同 `messageRef` 的临时流式消息。
    case agentMessageError(IMAgentMessageErrorEvent)
    /// @Agent 执行失败；仅发起人 personal 频道可见，不在群里留失败消息。
    case aiError(agentName: String, reason: String)
    /// Agent 建议把当前请求转为独立任务；仅作轻量个人提示。
    case aiSuggestTask(conversationId: String?, messageId: Int?, agentName: String)
    /// 本期未处理的事件类型（保留 type 便于日志/后续扩展）。
    case unknown(type: String)
}

struct IMAgentMessageStreamEvent: Decodable, Sendable, Equatable {
    let conversationId: String
    let messageRef: String
    let agentSessionRef: String
    let senderId: String
    let senderName: String
    let senderAvatar: String
    let delta: String
    let streamSeq: Int
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case messageRef = "message_ref"
        case agentSessionRef = "agent_session_ref"
        case senderId = "sender_id"
        case senderName = "sender_name"
        case senderAvatar = "sender_avatar"
        case delta
        case streamSeq = "stream_seq"
        case createdAt = "created_at"
    }
}

struct IMAgentMessageFinalEvent: Decodable, Sendable, Equatable {
    let conversationId: String
    let messageRef: String
    let agentSessionRef: String
    let senderId: String
    let senderName: String
    let senderAvatar: String
    let content: String
    let messageType: Int
    let metadata: IMMessageMetadata?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case messageRef = "message_ref"
        case agentSessionRef = "agent_session_ref"
        case senderId = "sender_id"
        case senderName = "sender_name"
        case senderAvatar = "sender_avatar"
        case content
        case messageType = "message_type"
        case metadata
        case createdAt = "created_at"
    }
}

struct IMAgentMessageErrorEvent: Decodable, Sendable, Equatable {
    let conversationId: String
    let messageRef: String
    let agentSessionRef: String
    let senderId: String
    let senderName: String
    let senderAvatar: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case messageRef = "message_ref"
        case agentSessionRef = "agent_session_ref"
        case senderId = "sender_id"
        case senderName = "sender_name"
        case senderAvatar = "sender_avatar"
    }
}

struct IMConversationPreviewUpdate: Decodable, Sendable, Equatable {
    let conversationId: String
    let organizationId: String
    let directoryScopeId: String?
    let messageId: Int
    let messageSeq: Int
    let preview: String
    let lastMessageAt: String?

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case organizationId = "organization_id"
        case directoryScopeId = "directory_scope_id"
        case messageId = "message_id"
        case messageSeq = "message_seq"
        case lastMessageAt = "last_message_at"
        case preview
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        conversationId = try c.decode(String.self, forKey: .conversationId)
        organizationId = try c.decodeIfPresent(String.self, forKey: .organizationId) ?? ""
        directoryScopeId = try c.decodeIfPresent(String.self, forKey: .directoryScopeId)
        messageId = try c.decodeIfPresent(Int.self, forKey: .messageId) ?? 0
        messageSeq = try c.decodeIfPresent(Int.self, forKey: .messageSeq) ?? 0
        preview = try c.decodeIfPresent(String.self, forKey: .preview) ?? ""
        lastMessageAt = try c.decodeIfPresent(String.self, forKey: .lastMessageAt)
    }
}

/// `im.read.receipt` 负载（`message_service` L1686）。
struct IMReadReceiptEvent: Decodable, Sendable, Equatable {
    let conversationId: String
    let userId: String
    let lastReadMessageId: Int?
    let lastReadSeq: Int

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case userId = "user_id"
        case lastReadMessageId = "last_read_message_id"
        case lastReadSeq = "last_read_seq"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        conversationId = try c.decodeIfPresent(String.self, forKey: .conversationId) ?? ""
        userId = try c.decodeIfPresent(String.self, forKey: .userId) ?? ""
        lastReadMessageId = try? c.decodeIfPresent(Int.self, forKey: .lastReadMessageId)
        lastReadSeq = try c.decodeIfPresent(Int.self, forKey: .lastReadSeq) ?? 0
    }
}

/// `im.reaction.added` / `im.reaction.removed` 负载（`message_service` L1931）。
struct IMReactionEvent: Decodable, Sendable, Equatable {
    let messageId: Int
    let userId: String
    let emoji: String

    enum CodingKeys: String, CodingKey {
        case messageId = "message_id"
        case userId = "user_id"
        case emoji
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        messageId = try c.decodeIfPresent(Int.self, forKey: .messageId) ?? 0
        userId = try c.decodeIfPresent(String.self, forKey: .userId) ?? ""
        emoji = try c.decodeIfPresent(String.self, forKey: .emoji) ?? ""
    }
}

/// `im.message.deleted` 负载（`message_service` L740）。
struct IMMessageDeletedEvent: Decodable, Sendable, Equatable {
    let messageId: Int

    enum CodingKeys: String, CodingKey {
        case messageId = "message_id"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        messageId = try c.decodeIfPresent(Int.self, forKey: .messageId) ?? 0
    }
}

private struct IMAIErrorEvent: Decodable {
    let agentName: String
    let reason: String

    enum CodingKeys: String, CodingKey {
        case agentName = "agent_name"
        case reason
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        agentName = try c.decodeIfPresent(String.self, forKey: .agentName) ?? ""
        reason = try c.decodeIfPresent(String.self, forKey: .reason) ?? ""
    }
}

private struct IMAISuggestTaskEvent: Decodable {
    let conversationId: String?
    let messageId: Int?
    let agentName: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case messageId = "message_id"
        case agentName = "agent_name"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        conversationId = try c.decodeIfPresent(String.self, forKey: .conversationId)
        messageId = try c.decodeIfPresent(Int.self, forKey: .messageId)
        agentName = try c.decodeIfPresent(String.self, forKey: .agentName) ?? ""
    }
}

/// `im.unread.update` 负载：两种形状共用同一 type——
/// 1) 新消息：带 `message_id` / `preview` 等（personal_base + mention）
/// 2) 已读回写：带 `marked_read` / `last_read_seq`（mark_as_read 后推给本人）
struct IMUnreadUpdate: Decodable, Sendable, Equatable {
    let conversationId: String
    let organizationId: String
    let directoryScopeId: String?
    let messageId: Int
    let messageSeq: Int
    let senderId: String
    let senderName: String
    let preview: String
    let lastMessageAt: String?
    let mention: Bool
    /// 非 nil 表示这是「已读回写」事件（应把该会话未读清零），与新消息事件互斥。
    let markedRead: Int?

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case organizationId = "organization_id"
        case directoryScopeId = "directory_scope_id"
        case messageId = "message_id"
        case messageSeq = "message_seq"
        case senderId = "sender_id"
        case senderName = "sender_name"
        case preview
        case lastMessageAt = "last_message_at"
        case mention
        case markedRead = "marked_read"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        conversationId = try c.decode(String.self, forKey: .conversationId)
        organizationId = try c.decodeIfPresent(String.self, forKey: .organizationId) ?? ""
        directoryScopeId = try c.decodeIfPresent(String.self, forKey: .directoryScopeId)
        messageId = try c.decodeIfPresent(Int.self, forKey: .messageId) ?? 0
        messageSeq = try c.decodeIfPresent(Int.self, forKey: .messageSeq) ?? 0
        senderId = try c.decodeIfPresent(String.self, forKey: .senderId) ?? ""
        senderName = try c.decodeIfPresent(String.self, forKey: .senderName) ?? ""
        preview = try c.decodeIfPresent(String.self, forKey: .preview) ?? ""
        lastMessageAt = try c.decodeIfPresent(String.self, forKey: .lastMessageAt)
        mention = try c.decodeIfPresent(Bool.self, forKey: .mention) ?? false
        markedRead = try? c.decodeIfPresent(Int.self, forKey: .markedRead)
    }

    /// 已读回写事件（应清零列表未读）。
    var isMarkedReadEvent: Bool { markedRead != nil }
}

/// `im.user.profile.updated` 负载。Django 会推给仍共享可见会话的参与者。
struct IMUserProfileUpdated: Decodable, Sendable, Equatable {
    let userId: String
    let nickname: String
    let username: String
    let avatar: String
    let avatarVersion: String
    let revision: Int

    enum CodingKeys: String, CodingKey {
        case userId = "id"
        case nickname, username, avatar, revision
        case avatarVersion = "avatar_version"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        userId = try c.decodeIfPresent(String.self, forKey: .userId) ?? ""
        nickname = try c.decodeIfPresent(String.self, forKey: .nickname) ?? ""
        username = try c.decodeIfPresent(String.self, forKey: .username) ?? ""
        avatar = try c.decodeIfPresent(String.self, forKey: .avatar) ?? ""
        avatarVersion = try c.decodeIfPresent(String.self, forKey: .avatarVersion) ?? ""
        revision = try c.decodeIfPresent(Int.self, forKey: .revision) ?? 0
    }

    var displayName: String {
        let preferred = nickname.trimmingCharacters(in: .whitespacesAndNewlines)
        return preferred.isEmpty
            ? username.trimmingCharacters(in: .whitespacesAndNewlines)
            : preferred
    }
}

/// Centrifugo publication 原始字节 → 类型化事件。
///
/// 外层信封形如 `{ "type": "im.*", "event_id": "...", "data": {...} }`；`im.typing` 例外，
/// 字段直接铺在顶层：`{ "type": "im.typing", "user_id": "..." }`。无法识别或解析失败返回
/// `nil`（调用方据此丢弃脏包，不影响连接）。
enum IMEventDecoder {
    private struct Head: Decodable {
        let type: String
    }

    private struct DataEnvelope<T: Decodable>: Decodable {
        let data: T
    }

    private struct TypingEnvelope: Decodable {
        let userId: String
        enum CodingKeys: String, CodingKey {
            case userId = "user_id"
        }
    }

    private struct HandoffUpdate: Decodable {
        let handoffId: String
        enum CodingKeys: String, CodingKey { case handoffId = "handoff_id" }
    }

    private struct SessionShareUpdate: Decodable {
        let shareId: String
        enum CodingKeys: String, CodingKey { case shareId = "share_id" }
    }

    private struct ConversationLabelsUpdated: Decodable {
        let conversationId: String
        let labels: [IMConversationLabel]

        enum CodingKeys: String, CodingKey {
            case conversationId = "conversation_id"
            case labels
        }
    }

    static func decode(_ raw: Data) -> IMRealtimeEvent? {
        let decoder = JSONDecoder()
        guard let head = try? decoder.decode(Head.self, from: raw) else { return nil }
        switch head.type {
        case "im.message":
            guard let env = try? decoder.decode(DataEnvelope<IMMessage>.self, from: raw) else { return nil }
            return .message(env.data)
        case "im.message.edited":
            guard let env = try? decoder.decode(DataEnvelope<IMMessage>.self, from: raw) else { return nil }
            return .messageEdited(env.data)
        case "im.message.deleted":
            guard let env = try? decoder.decode(DataEnvelope<IMMessageDeletedEvent>.self, from: raw) else { return nil }
            return .messageDeleted(messageId: env.data.messageId)
        case "im.message.pinned":
            guard let env = try? decoder.decode(DataEnvelope<IMMessage>.self, from: raw) else { return nil }
            return .messagePinned(env.data)
        case "im.message.unpinned":
            guard let env = try? decoder.decode(DataEnvelope<IMMessageDeletedEvent>.self, from: raw) else { return nil }
            return .messageUnpinned(messageId: env.data.messageId)
        case "im.reaction.added", "im.reaction.removed":
            guard let env = try? decoder.decode(DataEnvelope<IMReactionEvent>.self, from: raw) else { return nil }
            return .reaction(
                messageId: env.data.messageId,
                userId: env.data.userId,
                emoji: env.data.emoji,
                added: head.type == "im.reaction.added"
            )
        case "im.read.receipt":
            guard let env = try? decoder.decode(DataEnvelope<IMReadReceiptEvent>.self, from: raw) else { return nil }
            return .readReceipt(env.data)
        case "im.unread.update":
            guard let env = try? decoder.decode(DataEnvelope<IMUnreadUpdate>.self, from: raw) else { return nil }
            return .unreadUpdate(env.data)
        case "im.conversation.new":
            guard let env = try? decoder.decode(DataEnvelope<IMConversation>.self, from: raw) else { return nil }
            return .conversationNew(env.data)
        case "im.conversation.preview.updated":
            guard let env = try? decoder.decode(DataEnvelope<IMConversationPreviewUpdate>.self, from: raw) else {
                return nil
            }
            return .conversationPreviewUpdated(env.data)
        case "im.conversation.labels.updated":
            guard let env = try? decoder.decode(DataEnvelope<ConversationLabelsUpdated>.self, from: raw) else {
                return nil
            }
            return .conversationLabelsUpdated(
                conversationId: env.data.conversationId,
                labels: env.data.labels
            )
        case "im.user.profile.updated":
            guard let env = try? decoder.decode(DataEnvelope<IMUserProfileUpdated>.self, from: raw) else {
                return nil
            }
            return .userProfileUpdated(env.data)
        case "im.conversation.updated", "im.member.joined", "im.member.left":
            return .conversationChanged
        case "im.typing":
            guard let env = try? decoder.decode(TypingEnvelope.self, from: raw) else { return nil }
            return .typing(userId: env.userId)
        case "im.handoff.update":
            guard let env = try? decoder.decode(DataEnvelope<HandoffUpdate>.self, from: raw) else { return nil }
            return .handoffUpdate(handoffId: env.data.handoffId)
        case "im.session_share.update":
            guard let env = try? decoder.decode(DataEnvelope<SessionShareUpdate>.self, from: raw) else { return nil }
            return .sessionShareUpdate(shareId: env.data.shareId)
        case "im.agent.message.stream":
            guard let env = try? decoder.decode(DataEnvelope<IMAgentMessageStreamEvent>.self, from: raw) else {
                return nil
            }
            return .agentMessageStream(env.data)
        case "im.agent.message.final":
            guard let env = try? decoder.decode(DataEnvelope<IMAgentMessageFinalEvent>.self, from: raw) else {
                return nil
            }
            return .agentMessageFinal(env.data)
        case "im.agent.message.error":
            guard let env = try? decoder.decode(DataEnvelope<IMAgentMessageErrorEvent>.self, from: raw) else {
                return nil
            }
            return .agentMessageError(env.data)
        case "im.ai.error":
            guard let env = try? decoder.decode(DataEnvelope<IMAIErrorEvent>.self, from: raw) else { return nil }
            return .aiError(agentName: env.data.agentName, reason: env.data.reason)
        case "im.ai.suggest_task":
            guard let env = try? decoder.decode(DataEnvelope<IMAISuggestTaskEvent>.self, from: raw) else { return nil }
            return .aiSuggestTask(
                conversationId: env.data.conversationId,
                messageId: env.data.messageId,
                agentName: env.data.agentName
            )
        default:
            return .unknown(type: head.type)
        }
    }
}
